File: src/Mac/Sources/ClaudeQuotaCore/UsageParser.swift
import Foundation

/// Parses a `control_response` envelope into a `UsageSnapshot`.
///
/// The Windows parser (`src/Windows/Data/UsageParser.cs`) never hardcodes a path: it
/// depth-first-searches the tree for a key *containing* `five_hour` / `seven_day`, because the
/// container's shape is undocumented and has moved before. Phase 1 (docs/mac-port.md §3) found
/// that this is now one server-side key reorder away from reading the wrong bucket:
///
/// ```
///  0 five_hour   1 seven_day   2 seven_day_oauth_apps   3 seven_day_opus   4 seven_day_sonnet
///  … 7 limits    8 seven_day_cowork   9 seven_day_omelette   … 22 seven_day_breakdown
/// ```
///
/// `seven_day_oauth_apps`, `seven_day_cowork`, `seven_day_omelette` and `seven_day_breakdown`
/// carry no `opus`/`sonnet`/`haiku` qualifier, so "first unqualified match" is saved only by wire
/// order. Reading the wrong bucket is exactly a "confident wrong number".
///
/// Strategy, after the 2026-09-19 adversarial review:
///
///  1. **`limits[]` by `kind`** — self-describing; `session` / `weekly_all` matched `five_hour` /
///     `seven_day` bit-for-bit in every observation, microseconds included.
///  2. **The exact keys** `five_hour` and `seven_day`.
///  3. **Cross-check.** When both representations are present and usable they must agree. They
///     are two encodings of one number, so a disagreement means the response is in a state this
///     parser does not understand — during a server migration, say. That returns *no snapshot*
///     with an explanatory error, because the project's rule is that silence beats a wrong
///     verdict (docs/forecast-and-states.md, "Refusal → Measuring").
///  4. **The substring search**, last, and only when it is **unambiguous**. If several unqualified
///     `seven_day*` candidates are populated there is no principled way to choose between them --
///     `JSONSerialization` does not even preserve wire order -- so it reports unavailable rather
///     than picking one and attaching a caveat nobody downstream is obliged to read.
public enum UsageParser {
    public struct ParseResult: Sendable {
        public let snapshot: UsageSnapshot?
        public let error: String?
    }

    /// `controlResponse` is the decoded root envelope: `{"type":"control_response","response":{…}}`.
    public static func parse(controlResponse root: Any, now: Date = Date()) -> ParseResult {
        guard let rootObject = root as? [String: Any],
              let outer = rootObject["response"] as? [String: Any] else {
            return ParseResult(snapshot: nil, error: "control_response had no 'response' object")
        }
        // The payload is nested one level further: response.response holds rate_limits.
        let payload = (outer["response"] as? [String: Any]) ?? outer
        let subscriptionType = findSubscriptionType(payload)

        guard let rateLimits = payload["rate_limits"] as? [String: Any] else {
            // A logged-out or empty config dir answers `rate_limits: null` with
            // `rate_limits_available: false` (docs/mac-port.md §2). That is a legitimate,
            // expected state -- not a parse failure -- and must read as "no data", never as 0 %.
            let available = boolean(payload["rate_limits_available"])
            return ParseResult(
                snapshot: nil,
                error: available == false
                    ? "no rate_limits: this config directory is not logged in"
                    : "no rate_limits object in the response")
        }

        let fromLimits = readLimitsArray(rateLimits)
        let fromExact = readExactKeys(rateLimits)

        if let fromLimits, let fromExact, let disagreement = disagreement(fromLimits, fromExact) {
            return ParseResult(
                snapshot: nil,
                error: "limits[] and the exact keys disagree (\(disagreement)); refusing to guess")
        }

        // Prefer limits[] but fill anything it omits from the exact keys. They have already been
        // cross-checked, so this cannot import a contradicting value -- and dropping a deadline
        // that one encoding carries would make a live window look closed, which costs the
        // forecast its clock.
        if var reading = fromLimits {
            if let fromExact { reading = reading.completed(from: fromExact) }
            return ParseResult(snapshot: reading.snapshot(subscriptionType: subscriptionType, now: now),
                               error: nil)
        }
        if let fromExact {
            return ParseResult(snapshot: fromExact.snapshot(subscriptionType: subscriptionType, now: now),
                               error: nil)
        }

        switch readSubstringSearch(payload) {
        case .found(let reading):
            return ParseResult(
                snapshot: reading.snapshot(subscriptionType: subscriptionType, now: now),
                error: "read via the substring fallback: neither limits[] nor the exact keys were usable")
        case .ambiguous(let keys):
            return ParseResult(
                snapshot: nil,
                error: "ambiguous weekly buckets (\(keys.sorted().joined(separator: ", "))); refusing to guess")
        case .none:
            return ParseResult(snapshot: nil, error: "no usable five_hour/session node in the response")
        }
    }

    public static func parse(jsonLine: String, now: Date = Date()) -> ParseResult {
        guard let data = jsonLine.data(using: .utf8),
              let object = try? JSONSerialization.jsonObject(with: data) else {
            return ParseResult(snapshot: nil, error: "line is not valid JSON")
        }
        return parse(controlResponse: object, now: now)
    }

    // MARK: - One window's worth of readings

    struct Reading {
        var sessionPct: Double
        /// Optional: an expired 5 h window reports `resets_at: null` with `is_active: false`,
        /// which is a real state ("nothing running"), not a malformed reply.
        var sessionResetsAt: String?
        var sessionIsActive: Bool
        var weeklyPct: Double?
        var weeklyResetsAt: String?
        var weeklyIsActive: Bool
        var source: UsageSnapshot.Source

        /// Fills gaps from another reading of the same response. A reset present in either
        /// encoding means the window is open, whatever the other one omitted.
        func completed(from other: Reading) -> Reading {
            var merged = self
            if merged.sessionResetsAt == nil, let resets = other.sessionResetsAt {
                merged.sessionResetsAt = resets
                merged.sessionIsActive = true
            }
            if merged.weeklyPct == nil { merged.weeklyPct = other.weeklyPct }
            if merged.weeklyResetsAt == nil, let resets = other.weeklyResetsAt {
                merged.weeklyResetsAt = resets
                merged.weeklyIsActive = true
            }
            return merged
        }

        func snapshot(subscriptionType: String?, now: Date) -> UsageSnapshot {
            UsageSnapshot(
                sessionUtilization: sessionPct,
                sessionResetsAt: sessionResetsAt,
                weeklyUtilization: weeklyPct,
                weeklyResetsAt: weeklyResetsAt,
                subscriptionType: subscriptionType,
                sessionIsActive: sessionIsActive,
                weeklyIsActive: weeklyIsActive,
                observedAt: now,
                source: source)
        }
    }

    /// Two encodings of one number. A difference of more than half a point, or a different reset
    /// fingerprint, means they are not describing the same thing.
    static func disagreement(_ a: Reading, _ b: Reading) -> String? {
        if abs(a.sessionPct - b.sessionPct) > 0.5 {
            return "session \(a.sessionPct) vs \(b.sessionPct)"
        }
        // Only comparable when both carry one; an inactive window has none on either side.
        if let ar = a.sessionResetsAt, let br = b.sessionResetsAt, ar != br {
            return "session resets_at \(ar) vs \(br)"
        }
        if let aw = a.weeklyPct, let bw = b.weeklyPct, abs(aw - bw) > 0.5 {
            return "weekly \(aw) vs \(bw)"
        }
        if let ar = a.weeklyResetsAt, let br = b.weeklyResetsAt, ar != br {
            return "weekly resets_at \(ar) vs \(br)"
        }
        return nil
    }

    // MARK: - Strategy 1: the self-describing limits[] array

    static func readLimitsArray(_ rateLimits: [String: Any]) -> Reading? {
        guard let limits = rateLimits["limits"] as? [[String: Any]] else { return nil }
        var session: [String: Any]?
        var weekly: [String: Any]?
        // A duplicate `kind` means the response is malformed or ambiguous; taking the first and
        // ignoring the rest would be a silent choice between two different numbers.
        for entry in limits {
            switch entry["kind"] as? String {
            case "session":
                if session != nil { return nil }
                session = entry
            case "weekly_all":
                if weekly != nil { return nil }
                weekly = entry
            default: break   // weekly_scoped is a per-model bucket, never the "all models" ring
            }
        }
        guard let session, let sessionPct = validPercentage(session["percent"]) else { return nil }
        let sessionResets = resetString(session["resets_at"])
        // `is_active` is the server's own word for it; fall back to "has a deadline" when absent.
        let sessionActive = boolean(session["is_active"]) ?? (sessionResets != nil)

        // Weekly is validated independently: an invalid weekly must not invalidate a good
        // session reading, but it must not travel inside the snapshot either.
        var weeklyPct: Double?
        var weeklyResets: String?
        var weeklyActive = false
        if let weekly, let pct = validPercentage(weekly["percent"]) {
            weeklyPct = pct
            weeklyResets = resetString(weekly["resets_at"])
            weeklyActive = boolean(weekly["is_active"]) ?? (weeklyResets != nil)
        }
        return Reading(sessionPct: sessionPct, sessionResetsAt: sessionResets,
                       sessionIsActive: sessionActive,
                       weeklyPct: weeklyPct, weeklyResetsAt: weeklyResets,
                       weeklyIsActive: weeklyActive, source: .limitsArray)
    }

    // MARK: - Strategy 2: the exact keys

    static func readExactKeys(_ rateLimits: [String: Any]) -> Reading? {
        guard let fiveHour = rateLimits["five_hour"] as? [String: Any],
              let sessionPct = validPercentage(fiveHour["utilization"]) else { return nil }
        let sessionResets = resetString(fiveHour["resets_at"])

        var weeklyPct: Double?
        var weeklyResets: String?
        if let sevenDay = rateLimits["seven_day"] as? [String: Any],
           let pct = validPercentage(sevenDay["utilization"]) {
            weeklyPct = pct
            weeklyResets = resetString(sevenDay["resets_at"])
        }
        return Reading(sessionPct: sessionPct, sessionResetsAt: sessionResets,
                       sessionIsActive: sessionResets != nil,
                       weeklyPct: weeklyPct, weeklyResetsAt: weeklyResets,
                       weeklyIsActive: weeklyResets != nil, source: .exactKeys)
    }

    // MARK: - Strategy 3: the substring search, and only when unambiguous

    enum SubstringOutcome {
        case found(Reading)
        case ambiguous(Set<String>)
        case none
    }

    static func readSubstringSearch(_ payload: [String: Any]) -> SubstringOutcome {
        let sessionMatches = collect(payload, containing: "five_hour")
            .filter { $0.value is [String: Any] }
        guard sessionMatches.count == 1,
              let fiveHour = sessionMatches[0].value as? [String: Any],
              let sessionPct = validPercentage(fiveHour["utilization"]) else {
            if sessionMatches.count > 1 {
                return .ambiguous(Set(sessionMatches.map(\.key)))
            }
            return .none
        }
        let sessionResets = resetString(fiveHour["resets_at"])

        let qualifiers = ["opus", "sonnet", "haiku"]
        let weeklyCandidates = collect(payload, containing: "seven_day").filter { match in
            guard match.value is [String: Any] else { return false }
            let key = match.key.lowercased()
            return !qualifiers.contains { key.contains($0) }
        }
        // `JSONSerialization` returns an unordered Dictionary, so the Windows parser's
        // "first hit in wire order" tie-break cannot be reproduced. With more than one live
        // candidate the honest answer is that we do not know which is the all-models bucket.
        if weeklyCandidates.count > 1 {
            return .ambiguous(Set(weeklyCandidates.map(\.key)))
        }
        var weeklyPct: Double?
        var weeklyResets: String?
        if let weekly = weeklyCandidates.first?.value as? [String: Any],
           let pct = validPercentage(weekly["utilization"]) {
            weeklyPct = pct
            weeklyResets = resetString(weekly["resets_at"])
        }
        return .found(Reading(sessionPct: sessionPct, sessionResetsAt: sessionResets,
                              sessionIsActive: sessionResets != nil,
                              weeklyPct: weeklyPct, weeklyResetsAt: weeklyResets,
                              weeklyIsActive: weeklyResets != nil,
                              source: .substringSearch))
    }

    // MARK: - Helpers

    static func findSubscriptionType(_ payload: [String: Any]) -> String? {
        if let direct = payload["subscription_type"] as? String { return direct }
        return collect(payload, containing: "subscription_type")
            .compactMap { $0.value as? String }.first
    }

    /// Preorder depth-first walk; collects every value whose owning key contains `needle`.
    /// Unbounded recursion is not a concern here: `JSONSerialization` rejects deeply nested
    /// input before this ever sees it (verified — a 5 000-deep array fails to decode).
    static func collect(_ value: Any, containing needle: String) -> [(key: String, value: Any)] {
        var results: [(key: String, value: Any)] = []
        func walk(_ node: Any) {
            if let object = node as? [String: Any] {
                for (key, child) in object {
                    if key.lowercased().contains(needle) { results.append((key, child)) }
                    walk(child)
                }
            } else if let array = node as? [Any] {
                for item in array { walk(item) }
            }
        }
        walk(value)
        return results
    }

    /// The same field arrives as a JSON int in some responses and a JSON float in others.
    ///
    /// JSON booleans must NOT pass: `JSONSerialization` bridges `true` to an `NSNumber` whose
    /// `doubleValue` is `1.0`, so `{"percent": true}` would otherwise render as "1 % used" --
    /// a confident wrong number from a malformed reply. Verified against Foundation, not assumed.
    static func number(_ value: Any?) -> Double? {
        guard let value else { return nil }
        if let number = value as? NSNumber {
            if CFGetTypeID(number) == CFBooleanGetTypeID() { return nil }
            return number.doubleValue
        }
        if let d = value as? Double { return d }
        if let i = value as? Int { return Double(i) }
        return nil
    }

    static func boolean(_ value: Any?) -> Bool? {
        guard let number = value as? NSNumber, CFGetTypeID(number) == CFBooleanGetTypeID() else { return nil }
        return number.boolValue
    }

    /// docs/forecast-and-states.md, round-1 decision 9: utilization must be finite and in
    /// [0, 100] at the boundary, or the window is treated as missing rather than as 0 %.
    static func validPercentage(_ value: Any?) -> Double? {
        guard let parsed = number(value), UsageSnapshot.isValidPercentage(parsed) else { return nil }
        return parsed
    }

    /// Kept as the **raw string** from the wire: the exact byte sequence is the snapshot
    /// fingerprint the model dedups on. It is bounds-checked as a timestamp but never
    /// reformatted -- reformatting would destroy the microsecond digits that carry the signal.
    static func resetString(_ value: Any?) -> String? {
        guard let text = value as? String, !text.isEmpty, text.count <= 64 else { return nil }
        return text
    }
}

File: src/Windows/Data/UsageParser.cs
using System.Linq;
using System.Text.Json;
using ClaudeStatusBar.Model;

namespace ClaudeStatusBar.Data;

/// <summary>
/// Parses a control_response envelope into a UsageSnapshot.
///
/// The rate_limits container's exact shape is undocumented and has moved before
/// (docs/mac-port.md, "the weekly-node heuristic is one key-reorder from wrong"): a response can
/// carry many `seven_day`-ish sibling keys (`seven_day_oauth_apps`, `seven_day_cowork`,
/// `seven_day_omelette`, `seven_day_breakdown`, ...), none of which carry an
/// opus/sonnet/haiku qualifier either -- so "first unqualified match" was saved only by wire
/// order, and one server-side key reorder away from reading the wrong bucket. That is exactly a
/// "confident wrong number".
///
/// Strategy, matching src/Mac/Sources/ClaudeQuotaCore/UsageParser.swift exactly (docs/mac-port.md
/// §"DIFFERS (latent bug, not yet firing)"):
///
///  1. **`limits[]` by `kind`** -- self-describing (`kind: "session"` / `kind: "weekly_all"`);
///     matches `five_hour` / `seven_day` bit-for-bit, including the microseconds, whenever present.
///  2. **The exact keys** `five_hour` and `seven_day` -- found by an exact (not substring) DFS
///     key match, so a sibling like `seven_day_cowork` can never be mistaken for `seven_day`.
///  3. **Cross-check.** When both 1 and 2 are present and usable they must agree (session
///     percentage within 0.5, matching resets_at fingerprint when both carry one; same for
///     weekly). A disagreement means the response is in a shape this parser does not understand --
///     refuse rather than guess (docs/forecast-and-states.md, "Refusal -> Measuring").
///  4. **The substring search**, last, and only when it is unambiguous -- unchanged in spirit
///     from before, but now refuses outright (no snapshot at all) rather than silently falling
///     back to a wire-order guess when more than one qualifier-less `seven_day*`/`five_hour*`
///     candidate is populated.
///
/// Any parse failure is reported as a null snapshot with an explanatory error, never an
/// exception: a parse failure must degrade the gauge visibly, not crash or restart-loop the app.
/// </summary>
public static class UsageParser
{
    public static UsageSnapshot? TryParse(JsonElement controlResponse, out string? error)
    {
        try
        {
            if (!controlResponse.TryGetProperty("response", out var response))
            {
                error = "control_response had no 'response' property";
                return null;
            }

            string? subscriptionType = ReadSubscriptionType(response);

            Reading? fromLimits = ReadLimitsArray(response);
            Reading? fromExact = ReadExactKeys(response);

            if (fromLimits is { } limitsReading && fromExact is { } exactReading)
            {
                string? disagreement = Disagreement(limitsReading, exactReading);
                if (disagreement is not null)
                {
                    error = $"limits[] and the exact keys disagree ({disagreement}); refusing to guess";
                    return null;
                }
            }

            // Prefer limits[] but fill anything it omits from the exact keys -- already
            // cross-checked above, so this cannot import a contradicting value, and dropping a
            // deadline that one encoding carries would make a live window look closed, which
            // costs the forecast its clock.
            if (fromLimits is { } reading)
            {
                if (fromExact is { } exact) reading = Complete(reading, exact);
                error = null;
                return reading.ToSnapshot(subscriptionType);
            }
            if (fromExact is { } onlyExact)
            {
                error = null;
                return onlyExact.ToSnapshot(subscriptionType);
            }

            return ReadSubstringSearch(response, subscriptionType, out error);
        }
        catch (Exception ex)
        {
            error = ex.Message;
            return null;
        }
    }

    // ---- one window's worth of readings ----

    readonly record struct Reading(
        double SessionPct, string? SessionResetsAt, bool SessionIsActive,
        double? WeeklyPct, string? WeeklyResetsAt, bool WeeklyIsActive)
    {
        public UsageSnapshot ToSnapshot(string? subscriptionType) => new(
            SessionPct, SessionResetsAt, WeeklyPct, WeeklyResetsAt, DateTimeOffset.UtcNow,
            subscriptionType, SessionIsActive, WeeklyIsActive);
    }

    /// <summary>Fills gaps in <paramref name="reading"/> from <paramref name="other"/> (the same
    /// response's other encoding). A reset present in either encoding means the window is open,
    /// whatever the other one omitted.</summary>
    static Reading Complete(Reading reading, Reading other)
    {
        Reading merged = reading;
        if (merged.SessionResetsAt is null && other.SessionResetsAt is not null)
            merged = merged with { SessionResetsAt = other.SessionResetsAt, SessionIsActive = true };
        if (merged.WeeklyPct is null)
            merged = merged with { WeeklyPct = other.WeeklyPct };
        if (merged.WeeklyResetsAt is null && other.WeeklyResetsAt is not null)
            merged = merged with { WeeklyResetsAt = other.WeeklyResetsAt, WeeklyIsActive = true };
        return merged;
    }

    /// <summary>Two encodings of one number. A difference of more than half a point, or a
    /// different reset fingerprint, means they are not describing the same thing.</summary>
    static string? Disagreement(Reading a, Reading b)
    {
        if (Math.Abs(a.SessionPct - b.SessionPct) > 0.5)
            return $"session {a.SessionPct} vs {b.SessionPct}";
        if (a.SessionResetsAt is not null && b.SessionResetsAt is not null && a.SessionResetsAt != b.SessionResetsAt)
            return $"session resets_at {a.SessionResetsAt} vs {b.SessionResetsAt}";
        if (a.WeeklyPct is { } aw && b.WeeklyPct is { } bw && Math.Abs(aw - bw) > 0.5)
            return $"weekly {aw} vs {bw}";
        if (a.WeeklyResetsAt is not null && b.WeeklyResetsAt is not null && a.WeeklyResetsAt != b.WeeklyResetsAt)
            return $"weekly resets_at {a.WeeklyResetsAt} vs {b.WeeklyResetsAt}";
        return null;
    }

    // ---- strategy 1: the self-describing limits[] array ----

    static Reading? ReadLimitsArray(JsonElement response)
    {
        if (!TryFindExactKey(response, "limits", out var limits) || limits.ValueKind != JsonValueKind.Array)
            return null;

        JsonElement? session = null;
        JsonElement? weekly = null;
        // A duplicate `kind` means the response is malformed or ambiguous; taking the first and
        // ignoring the rest would be a silent choice between two different numbers.
        foreach (var entry in limits.EnumerateArray())
        {
            if (entry.ValueKind != JsonValueKind.Object) continue;
            string? kind = entry.TryGetProperty("kind", out var k) && k.ValueKind == JsonValueKind.String
                ? k.GetString() : null;
            switch (kind)
            {
                case "session":
                    if (session is not null) return null;
                    session = entry;
                    break;
                case "weekly_all":
                    if (weekly is not null) return null;
                    weekly = entry;
                    break;
                default: break; // weekly_scoped is a per-model bucket, never the "all models" ring
            }
        }

        if (session is not { } sessionEntry || ValidPercentage(sessionEntry, "percent") is not { } sessionPct)
            return null;
        string? sessionResets = ResetString(sessionEntry, "resets_at");
        // is_active is the server's own word for it; fall back to "has a deadline" when absent.
        bool sessionActive = ReadBoolean(sessionEntry, "is_active") ?? sessionResets is not null;

        // Weekly is validated independently: an invalid weekly must not invalidate a good
        // session reading, but it must not travel inside the snapshot either.
        double? weeklyPct = null;
        string? weeklyResets = null;
        bool weeklyActive = false;
        if (weekly is { } weeklyEntry && ValidPercentage(weeklyEntry, "percent") is { } pct)
        {
            weeklyPct = pct;
            weeklyResets = ResetString(weeklyEntry, "resets_at");
            weeklyActive = ReadBoolean(weeklyEntry, "is_active") ?? weeklyResets is not null;
        }

        return new Reading(sessionPct, sessionResets, sessionActive, weeklyPct, weeklyResets, weeklyActive);
    }

    // ---- strategy 2: the exact keys ----

    static Reading? ReadExactKeys(JsonElement response)
    {
        if (!TryFindExactKey(response, "five_hour", out var fiveHour) || fiveHour.ValueKind != JsonValueKind.Object)
            return null;
        if (ValidPercentage(fiveHour, "utilization") is not { } sessionPct) return null;
        string? sessionResets = ResetString(fiveHour, "resets_at");

        double? weeklyPct = null;
        string? weeklyResets = null;
        if (TryFindExactKey(response, "seven_day", out var sevenDay) && sevenDay.ValueKind == JsonValueKind.Object
            && ValidPercentage(sevenDay, "utilization") is { } pct)
        {
            weeklyPct = pct;
            weeklyResets = ResetString(sevenDay, "resets_at");
        }

        return new Reading(sessionPct, sessionResets, sessionResets is not null,
            weeklyPct, weeklyResets, weeklyResets is not null);
    }

    // ---- strategy 3: the substring search, and only when unambiguous ----

    static UsageSnapshot? ReadSubstringSearch(JsonElement response, string? subscriptionType, out string? error)
    {
        var sessionMatches = new List<(string Key, JsonElement Value)>();
        CollectBySubstring(response, "five_hour", sessionMatches);
        List<(string Key, JsonElement Value)> sessionObjects = sessionMatches
            .Where(m => m.Value.ValueKind == JsonValueKind.Object).ToList();

        if (sessionObjects.Count > 1)
        {
            error = $"ambiguous session buckets ({string.Join(", ", sessionObjects.Select(m => m.Key).Distinct().OrderBy(k => k, StringComparer.Ordinal))}); refusing to guess";
            return null;
        }
        if (sessionObjects.Count == 0 || ValidPercentage(sessionObjects[0].Value, "utilization") is not { } sessionPct)
        {
            error = "no usable five_hour/session node in the response";
            return null;
        }
        string? sessionResets = ResetString(sessionObjects[0].Value, "resets_at");

        var weeklyMatches = new List<(string Key, JsonElement Value)>();
        CollectBySubstring(response, "seven_day", weeklyMatches);
        List<(string Key, JsonElement Value)> weeklyCandidates = weeklyMatches
            .Where(m => m.Value.ValueKind == JsonValueKind.Object && !ContainsModelQualifier(m.Key)).ToList();

        if (weeklyCandidates.Count > 1)
        {
            error = $"ambiguous weekly buckets ({string.Join(", ", weeklyCandidates.Select(m => m.Key).Distinct().OrderBy(k => k, StringComparer.Ordinal))}); refusing to guess";
            return null;
        }

        double? weeklyPct = null;
        string? weeklyResets = null;
        if (weeklyCandidates.Count == 1 && ValidPercentage(weeklyCandidates[0].Value, "utilization") is { } pct)
        {
            weeklyPct = pct;
            weeklyResets = ResetString(weeklyCandidates[0].Value, "resets_at");
        }

        error = "read via the substring fallback: neither limits[] nor the exact keys were usable";
        var reading = new Reading(sessionPct, sessionResets, sessionResets is not null,
            weeklyPct, weeklyResets, weeklyResets is not null);
        return reading.ToSnapshot(subscriptionType);
    }

    // ---- helpers ----

    /// <summary>
    /// docs/multi-account.md: get_usage carries subscription_type (max, team, ...) for the
    /// account that child is logged into -- used only for account labelling
    /// (Model/AccountLabel.cs), never for the quota model itself.
    /// </summary>
    static string? ReadSubscriptionType(JsonElement response)
    {
        var matches = new List<(string Key, JsonElement Value)>();
        CollectBySubstring(response, "subscription_type", matches);
        if (matches.Count == 0) return null;
        return matches[0].Value.ValueKind == JsonValueKind.String ? matches[0].Value.GetString() : null;
    }

    static bool ContainsModelQualifier(string key) =>
        key.Contains("opus", StringComparison.OrdinalIgnoreCase) ||
        key.Contains("sonnet", StringComparison.OrdinalIgnoreCase) ||
        key.Contains("haiku", StringComparison.OrdinalIgnoreCase);

    /// <summary>Preorder depth-first walk; collects every object whose owning key contains needle.</summary>
    static void CollectBySubstring(JsonElement element, string needle, List<(string Key, JsonElement Value)> results)
    {
        if (element.ValueKind == JsonValueKind.Object)
        {
            foreach (var prop in element.EnumerateObject())
            {
                if (prop.Name.Contains(needle, StringComparison.OrdinalIgnoreCase))
                    results.Add((prop.Name, prop.Value));
                CollectBySubstring(prop.Value, needle, results);
            }
        }
        else if (element.ValueKind == JsonValueKind.Array)
        {
            foreach (var item in element.EnumerateArray())
                CollectBySubstring(item, needle, results);
        }
    }

    /// <summary>Preorder depth-first walk for an EXACT (not substring) key match, so a sibling
    /// like `seven_day_cowork` can never be mistaken for `seven_day`. Returns the first hit in
    /// DFS/wire order, per the same "first hit" convention CollectBySubstring already used.</summary>
    static bool TryFindExactKey(JsonElement element, string name, out JsonElement result)
    {
        if (element.ValueKind == JsonValueKind.Object)
        {
            foreach (var prop in element.EnumerateObject())
            {
                if (string.Equals(prop.Name, name, StringComparison.OrdinalIgnoreCase))
                {
                    result = prop.Value;
                    return true;
                }
            }
            foreach (var prop in element.EnumerateObject())
            {
                if (TryFindExactKey(prop.Value, name, out result)) return true;
            }
        }
        else if (element.ValueKind == JsonValueKind.Array)
        {
            foreach (var item in element.EnumerateArray())
            {
                if (TryFindExactKey(item, name, out result)) return true;
            }
        }
        result = default;
        return false;
    }

    /// <summary>
    /// The same field arrives as a JSON int in some responses and a JSON float in others, always
    /// read with GetDouble(). A JSON boolean must NOT pass (System.Text.Json would otherwise
    /// coerce `true`/`false` into 1/0 via GetDouble() on some numeric-ish paths) -- a malformed
    /// `{"percent": true}` must never render as "1 % used", a confident wrong number.
    /// docs/forecast-and-states.md, round-1 decision 9: utilization must be finite and in
    /// [0, 100] at the boundary, or the window is treated as missing rather than as 0 %.
    /// </summary>
    static double? ValidPercentage(JsonElement obj, string prop)
    {
        if (!obj.TryGetProperty(prop, out var v) || v.ValueKind != JsonValueKind.Number) return null;
        double d = v.GetDouble();
        return QuotaTimeUtil.IsValidPct(d) ? d : null;
    }

    /// <summary>
    /// Kept as the raw string from the wire: the exact byte sequence is the snapshot fingerprint
    /// the model dedups on. It is bounds-checked as a timestamp elsewhere but never reformatted
    /// here -- reformatting would destroy the microsecond digits that carry the signal.
    /// </summary>
    static string? ResetString(JsonElement obj, string prop) =>
        obj.TryGetProperty(prop, out var v) && v.ValueKind == JsonValueKind.String ? v.GetString() : null;

    static bool? ReadBoolean(JsonElement obj, string prop) =>
        obj.TryGetProperty(prop, out var v) && (v.ValueKind == JsonValueKind.True || v.ValueKind == JsonValueKind.False)
            ? v.GetBoolean() : null;
}
